src/Microsoft.AspNet.Razor.Runtime/TagHelpers/ReadOnlyTagHelperAttributes.cs

```diff
@@ -0,0 +1,199 @@
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Razor.Runtime.TagHelpers
+{
+    /// <summary>
+    /// Represents a read-only collection of <typeparamref name="TAttribute"/>s.
+    /// </summary>
+    /// <typeparam name="TAttribute">
+    /// The type of <see cref="IReadOnlyTagHelperAttribute"/>s in the collection.
+    /// </typeparam>
+    public class ReadOnlyTagHelperAttributes<TAttribute> : IReadOnlyList<TAttribute>
+        where TAttribute : IReadOnlyTagHelperAttribute
+    {
+        /// <summary>
+        /// Instantiates a new instance of the <see cref="ReadOnlyTagHelperAttributes{TAttribute}"/> with an empty
+        /// collection.
+        /// </summary>
+        protected ReadOnlyTagHelperAttributes()
+        {
+            Attributes = new List<TAttribute>();
+        }
+
+        /// <summary>
+        /// Instantiates a new instance of the <see cref="ReadOnlyTagHelperAttributes{TAttribute}"/> with the specified
+        /// <paramref name="attributes"/>.
+        /// </summary>
+        /// <param name="attributes">The collection to wrap.</param>
+        public ReadOnlyTagHelperAttributes([NotNull] IEnumerable<TAttribute> attributes)
+        {
+            Attributes = new List<TAttribute>(attributes);
+        }
+
+        /// <summary>
+        /// The underlying collection of <typeparamref name="TAttribute"/>s.
+        /// </summary>
+        /// <remarks>Modifying this <see cref="List{TAttribute}"/>s affects what is exposed by this
+        /// <see cref="ReadOnlyTagHelperAttributes{TAttribute}"/> instance.</remarks>
+        protected List<TAttribute> Attributes { get; }
+
+        /// <inheritdoc />
+        public TAttribute this[int index]
+        {
+            get
+            {
+                return Attributes[index];
+            }
+        }
+
+        /// <summary>
+        /// Gets the first <typeparamref name="TAttribute"/> whose <typeparamref name="TAttribute.Name"/> matches the
+        /// specified <paramref name="name"/>.
+        /// </summary>
+        /// <param name="name">
+        /// The <typeparamref name="TAttribute.Name"/> of the <typeparamref name="TAttribute"/> to get.
+        /// </param>
+        /// <returns>The first <typeparamref name="TAttribute"/> whose <typeparamref name="TAttribute.Name"/> matches
+        /// the specified <paramref name="name"/>.
+        /// </returns>
+        /// <remarks><paramref name="name"/> is compared case-insensitively.</remarks>
+        public TAttribute this[[NotNull] string name]
+        {
+            get
+            {
+                return Attributes.FirstOrDefault(attribute => NameEquals(name, attribute));
+            }
+        }
+
+        /// <inheritdoc />
+        public int Count
+        {
+            get
+            {
+                return Attributes.Count;
+            }
+        }
+
+        /// <inheritdoc />
+        public bool IsReadOnly
```

Comment:
It's not lying, it's about what's exposed to the user. That'd be like doing:

```C#
var dict = new Dictionary<...>();
Console.WriteLine(dict.IsReadOnly); // false
IReadOnlyDictionary<...> readOnlyDict = dict;
Console.WriteLine(dict.IsReadOnly); // If it was virtual this would return false, however it returns true
```

Essentially I'm doing the same thing that the other read-only implementations do. Only difference is the others have an interface, we don't so `new` gets the same capability.